hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ExpiringCache.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.util;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A thread-safe cache with automatic expiration and cleanup.
+ */
+public class ExpiringCache <K, V> {
```

Comment:
ExpiringCache is a generic class. Can we replace it with a utility class such as CaffeineCache?

Reply:
It's a good suggestion. Caffeine is indeed a robust and high-performance caching library. However, since Hadoop already depends on Guava,  using Guava’s Cache instead of introducing a new dependency maybe a good choice. Guava provides a CacheBuilder and LoadingCache that can fulfill our needs for automatic expiration, cleanup, and concurrency.